add docs about istio sidecar (#14766)
kyma-project/kyma: docs/01-overview/main-areas/service-mesh/smsh-04-istio-sidecars-in-kyma.md

---
title: Istio sidecars in Kyma and why you want them
---

> Istio installed as part of Kyma has [numerous default configuration](./smsh-02-default-istio-setup-in-kyma.md) configured to have automatic sidecar injection enabled by default. It means that all pods of your workloads (Deployments, StatefulSets and other) will get their own sidecar proxy container running next to your application (provided your workload is not in `istio-system` or `kube-system` namespace). This has numerous benefits, mentioned below, that you get out of the box, but are complex to manage without service mesh.

Reviewer: ```suggestion
Istio installed as part of Kyma has [default configuration](./smsh-02-default-istio-setup-in-kyma.md) with automatic Istio proxy sidecar injection enabled by default. It means that all pods of your workloads (Deployments, StatefulSets and other) will get their own sidecar proxy container running next to your application (provided your workload is not in `istio-system` or `kube-system` namespace). This is a requirement for the resource to be a part of Istio service mesh. It has numerous benefits, mentioned below, that you get out of the box, but are complex to manage without this feature.
```

## Secure communication

> One of [Kyma's default configuration option](./smsh-02-default-istio-setup-in-kyma.md) is [Peer authentication](https://istio.io/latest/docs/concepts/security/#peer-authentication) set to `STRICT` mode cluster-wide. This ensures that your workload only accepts [mutual TLS traffic](https://www.cloudflare.com/learning/access-management/what-is-mutual-tls/) where both, client and server certificates are validated and traffic is encrypted. This provides each service with a strong identity and Istio provides a key management system that automates key and certificate management.

Reviewer: ```suggestion
One of [Kyma's default configuration option](./smsh-02-default-istio-setup-in-kyma.md) is [Peer authentication](https://istio.io/latest/docs/concepts/security/#peer-authentication) set to `STRICT` mode cluster-wide. This ensures that your workload only accepts [mutual TLS traffic](https://www.cloudflare.com/learning/access-management/what-is-mutual-tls/) where both, client and server certificates are validated having all traffic encrypted. This provides each service a strong identity with reliable key and certificate management system.
```

> Another security benefit of having sidecar proxy is possibility to perform request authentication for your service. Istio enables [Request Authentication](https://istio.io/latest/docs/reference/config/security/request_authentication/) using JSON Web Token (JWT) validation using a custom authentication provider. You can learn how to secure your service using JWT tokens you can follow [the tutorial](../../../03-tutorials/00-api-exposure/apix-05-expose-and-secure-workload-jwt.md)

Reviewer: ```suggestion
Another security benefit of having sidecar proxy is possibility to perform request authentication for your service. Istio enables [Request Authentication](https://istio.io/latest/docs/reference/config/security/request_authentication/) using JSON Web Token (JWT) validation using a custom authentication provider. You can learn how to secure your service using JWT in [this tutorial](../../../03-tutorials/00-api-exposure/apix-05-expose-and-secure-workload-jwt.md).
```

> Another security benefit of having sidecar proxy is possibility to perform request authentication for your service. Istio enables [Request Authentication](https://istio.io/latest/docs/reference/config/security/request_authentication/) using JSON Web Token (JWT) validation using a custom authentication provider. You can learn how to secure your service using JWT tokens you can follow [the tutorial](../../../03-tutorials/00-api-exposure/apix-05-expose-and-secure-workload-jwt.md)

Reviewer: Do we want to also mention Authorization Policies?

## Observability

> Another default configuration for Istio is related to tracing. Istio is performing a global tracing and is using Zipkin protocol to sent tracing data to the Kyma's tracing component. You can read more about the process in the [Tracing Architecture](../../../05-technical-reference/00-architecture/obsv-03-architecture-tracing.md) section.

Reviewer: ```suggestion
Another benefit of having Istio proxies is related to tracing. Istio is performing global tracing and forwarding acquired data to the [Kyma's tracing component](../../../01-overview/main-areas/service-mesh/smsh-04-istio-sidecars-in-kyma.md) with using [Zipkin protocol](https://zipkin.io). You can read more about the process in the [Tracing Architecture](../../../05-technical-reference/00-architecture/obsv-03-architecture-tracing.md) section.
```

> Another benefit of having Istio proxies is related to tracing. Istio is performing global tracing and forwarding acquired data to the [Kyma's tracing component](../../../01-overview/main-areas/service-mesh/smsh-04-istio-sidecars-in-kyma.md) with using [Zipkin protocol](https://zipkin.io). You can read more about the process in the [Tracing Architecture](../../../05-technical-reference/00-architecture/obsv-03-architecture-tracing.md) section.

Reviewer: ```suggestion
Another benefit of having Istio proxies is related to tracing. Istio is performing global tracing and forwarding acquired data to the [Kyma's tracing component](../../../01-overview/main-areas/observability/obsv-03-tracing-in-kyma.md) using [Zipkin protocol](https://zipkin.io). You can read more about the process in the [Tracing Architecture](../../../05-technical-reference/00-architecture/obsv-03-architecture-tracing.md) section.
```

Kiali is another tool that comes as [separate Kyma component](../../../05-technical-reference/00-architecture/obsv-04-architecture-kiali.md) and Kyma configures Istio to export metrics necessary to support Kiali features that facilitates managing, visualizing and troubleshooting your mesh.

> Moreover, Kyma provides [Istio specific dashboards](https://istio.io/latest/docs/ops/integrations/grafana/#configuration) for [Grafana (monitoring component)](../../../05-technical-reference/00-architecture/obsv-01-architecture-monitoring.md) and together with metrics exposed by the sidecar proxy provides better visibility into workloads and mesh control plane performance.

Reviewer: ```suggestion
Moreover, Kyma provides [Istio specific dashboards](https://istio.io/latest/docs/ops/integrations/grafana/#configuration) for [Grafana (monitoring component)](../../../05-technical-reference/00-architecture/obsv-01-architecture-monitoring.md) and together with metrics exposed by the Istio sidecar provides better visibility into workloads and mesh control plane performance.
```

> The sidecar proxy enables all the advanced observability, that would not be possible without instrumentation code within your application that could complicate your domain code.

Reviewer: ```suggestion
Being part of Istio service mesh enables all this advanced observability features, that would not be possible without advanced instrumentation code within your application.
```

[Traffic Shifting](https://istio.io/latest/docs/tasks/traffic-management/traffic-shifting/) and [Request Routing](https://istio.io/latest/docs/tasks/traffic-management/request-routing/) allows developers to release their software using techniques like canary releases and A/B testing to make release process faster and more reliable.

[Mirroring](https://istio.io/latest/docs/tasks/traffic-management/mirroring/) and [Fault Injection](https://istio.io/latest/docs/tasks/traffic-management/fault-injection/) can be used for testing and audit purposes to improve resiliency of your applications.

### Resiliency

Application resiliency is an important topic within traffic management functionalities. Traditionally resiliency features like timeouts, retries and circuit breakers were implemented by application libraries, however with service mesh you can delegate that tasks to the mesh and the same configuration options will work regardless of the programming language of your application. You can read more about it in [Network resilience and testing](https://istio.io/latest/docs/concepts/traffic-management/#network-resilience-and-testing)

> ## Sidecar proxy behavior during Kyma upgrade.

Reviewer 2: ```suggestion
## Sidecar proxy behavior during Kyma upgrade
```

> When upgrading Kyma we want to have full compatibility of existing workloads with the upgraded version of Istio. To achieve that we try to perform `rollout restart` of the workloads whenever possible because this ensures that newest version of sidecar proxy will be injected into the pods. There are exception when it's not possible to restart workloads and the detailed behaviour is described in [troubleshooting guide](https://kyma-project.io/docs/kyma/latest/04-operation-guides/troubleshooting/apix-09-upgrade-sidecar-proxy#cause) 

Reviewer: ```suggestion
When upgrading Kyma we want to have full compatibility of existing workloads with the upgraded version of Istio. To achieve that we try to perform `rollout restart` of the workloads whenever possible because this ensures that newest version of sidecar proxy will be injected into the pods. There are exceptions when it's not possible to restart workloads and this behaviour is described in [troubleshooting guide](https://kyma-project.io/docs/kyma/latest/04-operation-guides/troubleshooting/apix-09-upgrade-sidecar-proxy#cause) 
```

---
title: Istio sidecars in Kyma and why you want them
---

> 

Reviewer 2: In Netlify preview, it looked weird that this paragraph has no headline.

```suggestion

## Purpose of Istio sidecars

```

By default, Istio installed as part of Kyma [is configured](./smsh-02-default-istio-setup-in-kyma.md) with automatic Istio proxy sidecar injection enabled. This means that all Pods of your workloads (such as deployments and StatefulSets; except any workloads in the `istio-system` or `kube-system` Namespaces) get their own sidecar proxy container running next to your application. With an Istio sidecar, the resource becomes part of Istio service mesh, which brings the following benefits that would be complex to manage otherwise.

## Secure communication

In Kyma's [default Istio configuration](./smsh-02-default-istio-setup-in-kyma.md), [peer authentication](https://istio.io/latest/docs/concepts/security/#peer-authentication) is set to cluster-wide `STRICT` mode. This ensures that your workload only accepts [mutual TLS traffic](https://www.cloudflare.com/learning/access-management/what-is-mutual-tls/) where both, client and server certificates, are validated to have all traffic encrypted. This provides each service with a strong identity, with reliable key and certificate management system.

> Another security benefit of having a sidecar proxy is that you can perform request authentication for your service. Istio enables [request authentication](https://istio.io/latest/docs/reference/config/security/request_authentication/) with JSON Web Token (JWT) validation using a custom authentication provider. Learn how to [secure your service using JWT](../../../03-tutorials/00-api-exposure/apix-05-expose-and-secure-workload-jwt.md).

Reviewer 2: We might want to explain that at the first occurrence, not the second one :D 

```suggestion
Another security benefit of having a sidecar proxy is that you can perform [request authentication](https://istio.io/latest/docs/reference/config/security/request_authentication/) for your service. Istio enables request authentication with JSON Web Token (JWT) validation using a custom authentication provider. Learn how to [secure your workload using JWT](../../../03-tutorials/00-api-exposure/apix-05-expose-and-secure-workload-jwt.md).
```
Title of the link target says "workload", so I'd use that here, too.

## Observability

> Furthermore, Istio proxies improve tracing: Istio performs global tracing and forwards the data to the [Kyma's tracing component](../../../01-overview/main-areas/observability/obsv-03-tracing-in-kyma.md) using the [Zipkin protocol](https://zipkin.io). Learn more about the process in [Tracing Architecture](../../../05-technical-reference/00-architecture/obsv-03-architecture-tracing.md).

Reviewer 2: ```suggestion
Furthermore, Istio proxies improve tracing: Istio performs global tracing and forwards the data to [Kyma's tracing component](../../../01-overview/main-areas/observability/obsv-03-tracing-in-kyma.md) using the [Zipkin protocol](https://zipkin.io). Learn more about the process in [Tracing Architecture](../../../05-technical-reference/00-architecture/obsv-03-architecture-tracing.md).
```

> Kiali is another tool that comes as [separate Kyma component](../../../05-technical-reference/00-architecture/obsv-04-architecture-kiali.md), and Kyma configures Istio to export metrics necessary to support Kiali features that facilitate managing, visualizing, and troubleshooting your service mesh.

Reviewer 2: ```suggestion
Kiali is another tool that comes as [separate Kyma component](../../../05-technical-reference/00-architecture/obsv-04-architecture-kiali.md); and Kyma configures Istio to export metrics necessary to support Kiali features that facilitate managing, visualizing, and troubleshooting your service mesh.
```

title: Istio sidecars in Kyma and why you want them
---

> By default, Istio installed as part of Kyma [is configured](./smsh-02-default-istio-setup-in-kyma.md) with automatic Istio proxy sidecar injection enabled. This means that all Pods of your workloads (such as deployments and StatefulSets; except any workloads in the `istio-system` or `kube-system` Namespaces) get their own sidecar proxy container running next to your application. With an Istio sidecar, the resource becomes part of Istio service mesh, which brings the following benefits that would be complex to manage otherwise.

Reviewer 2: Isn’t that what we plan to change - it’s going to be **_disabled_** by default? 
Why do we say it's **_enabled_** by default?

Moreover, Kyma provides [Istio specific dashboards](https://istio.io/latest/docs/ops/integrations/grafana/#configuration) for [Grafana (monitoring component)](../../../05-technical-reference/00-architecture/obsv-01-architecture-monitoring.md) and together with metrics exposed by the sidecar proxy provides better visibility into workloads and mesh control plane performance.

> Being part of Istio service mesh enables all this advanced observability features, that would not be possible without advanced instrumentation code within your application.

Reviewer 2: ```suggestion
Being part of Istio service mesh enables all these advanced observability features, which would not be possible without advanced instrumentation code within your application.
```
